Sources/Purchasing/IntroEligibilityCalculator.swift

```diff
@@ -86,28 +96,44 @@ extension IntroEligibilityCalculator: @unchecked Sendable {}
 @available(iOS 12.0, macOS 10.14, macCatalyst 13.0, tvOS 12.0, watchOS 6.2, *)
 private extension IntroEligibilityCalculator {
 
-    func checkEligibility(candidateProducts: Set<StoreProduct>,
-                          purchasedProductsWithIntroOffers: Set<StoreProduct>) -> [String: IntroEligibilityStatus] {
+    func checkEligibility(
+        candidateProducts: Set<StoreProduct>,
+        activeSubscriptionsProducts: Set<StoreProduct>,
+        expiredSubscriptionProducts: Set<StoreProduct>,
+        expiredTrialProducts: Set<StoreProduct>
+    ) -> [String: IntroEligibilityStatus] {
         var result: [String: IntroEligibilityStatus] = [:]
 
         for candidate in candidateProducts {
             guard candidate.subscriptionPeriod != nil else {
-                result[candidate.productIdentifier] = IntroEligibilityStatus.unknown
+                result[candidate.productIdentifier] = .unknown
                 continue
             }
-            let usedIntroForProductIdentifier = purchasedProductsWithIntroOffers
-                .contains { purchased in
-                    let foundByGroupId = (candidate.subscriptionGroupIdentifier != nil
-                        && candidate.subscriptionGroupIdentifier == purchased.subscriptionGroupIdentifier)
-                    return foundByGroupId
+            let activeSubscriptionInGroup = (
+                candidate.subscriptionGroupIdentifier != nil &&
+                activeSubscriptionsProducts.contains {
+                    $0.subscriptionGroupIdentifier == candidate.subscriptionGroupIdentifier
                 }
+            )
+            let expiredSubscriptionForProduct = expiredSubscriptionProducts
+                .lazy
+                .map(\.productIdentifier)
+                .contains(candidate.productIdentifier)
```

Comment:
We could potentially have a false negative here:
Expired subs for the product only matter if you have used a free trial for the product. 

For example, if you subscribe to `yearly 1`, and it has no free trial available: 
- you're not eligible while the subscription to `yearly 1` is active
- if the developer then _adds_ a free trial to `yearly 1` and you cancel the subscription, you're eligible because you haven't used a free trial for this product (or its group), and one is available

Reply:
<img width="669" alt="image" src="https://user-images.githubusercontent.com/3922667/212907290-7b46e72e-c71d-4e61-bdcb-da01c09bbf29.png">

https://developer.apple.com/documentation/storekit/in-app_purchase/original_api_for_in-app_purchase/subscriptions_and_offers/implementing_introductory_offers_in_your_app

Reply:
Thanks a lot for the thorough review!

I think my TDD approach has kinda failed me here. I’ve been writing tests and finding the correct implementation that satisfies all tests, but it led me to the wrong implementation.

I misinterpreted the `is_in_trial_period` property in the receipt, I figured that wouldn't be `true` for an expired subscription, but now I see that it is (and verified through integration tests).
So your suggested implementation works after I fix the incorrect unit test 👍🏻 

I've also add a new unit test to cover this false positive you described, which failed before the fix:
```swift
func testCheckTrialOrIntroDiscountEligibilityReturnsEligibleForPreviouslyOwnedSubscriptionWithUnusedTrial() {
    self.testEligibility(
        purchaseExpirationsByProductIdentifier: [
            ("com.revenuecat.product1", Date().addingTimeInterval(-1000), false)
        ],
        productsInGroups: [
            "com.revenuecat.product1": (groupID: "group1", hasTrial: true)
        ],
        expectedResult: [
            "com.revenuecat.product1": .eligible
        ]
    )
}
```